[ECF-1-101] Coordinator registration workflow

amazimbe opened the pull request:
### Context

Coordinator registration workflow and page content as depicted in these designs:
https://miro.com/app/board/o9J_lZ2k8oc=/
https://miro.com/app/board/o9J_lZ2k8oc=/

### Changes proposed in this pull request
Mostly rendering new pages based on the workflow.

Removing old pages that are no longer relevant

kicferk1 commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -1,45 +1,36 @@
 
   def create
     validate_creation_parameters
+    build_resource(sign_up_params)
```
What does `build_resource` do?

amazimbe replied:
It's like User.build(params) but it comes from devise. I had to override devise on this controller action so that I can render a template rather than a flash message.

kicferk1 commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -1,45 +1,36 @@
-          InductionCoordinatorProfile.create!(user: resource, schools: [@school])
-        end
-      end
+      resource.save!
```
Is this some lovely Rails magic, something like controller having a `resource` that it understands from controller name or something like that? Is it a devise thing?

amazimbe replied:
Yes, devise magic! Had to override devise as explained earlier

kicferk1 commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -49,23 +40,17 @@ def email_domain
-    else
-      redirect_to root_path, alert: "Someone from your school has already signed up"
-    end
+  def render_start_registration
+    flash.now[:alert] = "Your details did not match any schools."
+    render :start_registration
```
I don't know if we want to get into pattern of flash messages - do we render any validation on `start_registration`? Could we add an error there, that will show in error summary?

amazimbe replied:
No we are not rendering any validation errors at the moment. The form has school as well as  user information and we are running checks on the relationship between these two and rendering the correct template based on that. This one is the only error message I think the rest are just information pages. Our designer was quite adamant that I can't replace these pages with error / flash messages as the two are fundamentally serving different purposes.

kicferk1 replied:
Well, yeah, I agree with not removing pages.

What I was thinking about was showing the same `"Your details did not match any schools."` message in form of a form error, rather than a flash message. Do you think that would be possible?

amazimbe replied:
I think what you describe is better for model validations rather than when checking connections between models like we are doing here. I'm sure it can be done. One way would be to define a new entity that has a user and a school and does the checks. I just don't think it would be worth the cost.

kicferk1 replied:
Alright, flash message it is then :)

amazimbe replied:
@kicferk1 I did a demo of this with our designers and they will be splitting this into two screens so we can validate a single entity. That should address the error messaging concerns you had but it may be part of another pr.

kicferk1 replied:
Thanks for following up on it, sounds good to me to make it a separate PR.

kicferk1 commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -49,23 +40,17 @@ def email_domain
 
   def validate_creation_parameters
-    @school = School.find(params[:user][:school_id])
+    @school = School.find_by(id: params[:user][:school_id])
```
I thought those are identical. Are they not?

amazimbe replied:
find_by won't cause an exception if there is no record. It will just return nil instead.

kicferk1 commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -49,23 +40,17 @@ def email_domain
 
-    raise ActionController::BadRequest if @email.nil?
+    raise ActionController::BadRequest if @email.blank?
+    raise ActionController::BadRequest if @school.blank?
```
Do the bad requests get rendered in a user friendly way?

amazimbe replied:
I don't think they are, not in development anyway. Let me check a prod-like environment for that.

amazimbe replied:
@kicferk1 yes you get a nice looking page that something went wrong.

kicferk1 replied:
Alright, good enough for now, eventually would be nice to have some more specific pages describing the error and how to fix it, but not on this PR.

kicferk1 commented on app/views/users/confirmations/confirmed.html.erb:
```diff
@@ -0,0 +1,8 @@
+<div class="govuk-grid-row">
+  <h2 class="govuk-heading-m confirmed-banner">Your school is registered</h2>
```
Is this header gonna be full-width? Do we want to wrap it in a div saying so? 
```
<div class="govuk-grid-column-full-width">
```

Or should it be inside the two-thirds column?

amazimbe replied:
It will be full width

kicferk1 commented on app/views/users/registrations/school_partially_registered.html.erb:
```diff
@@ -0,0 +1,9 @@
+    <h2 class="govuk-heading-l">Your need to verify your email</h2>
+    <ul class="govuk-list govuk-list--bullet">
+      <li>check your email for our verification link (it may be in your spam or junk folder)</li>
+      <li>or <%= govuk_link_to "register again", users_new_path %> in 24 hours</li>
```
Out of curiosity, does the 24h part work at the moment?

amazimbe replied:
No, I have a trello ticket for it though

kicferk1 commented on app/views/users/registrations/start_registration.html.erb:
```diff
@@ -1,18 +1,22 @@
-    <%= form_for(:induction_coordinator_profile, url: users_check_email_path) do |f| %>
+    <h2 class="govuk-heading-l">Register your school</h2>
 
+    <%= form_for(:induction_coordinator_profile, url: users_check_details_path) do |f| %>
+      <%= f.govuk_text_field(
+            :school_urn,
+            value: @urn,
+            hint: { text: "6 digit number", class: "govuk-hint" },
+            label: { text: "Your school's Unique Reference Number (URN)", class: "govuk-label govuk-label--s" })
+      %>
+      <p>><%= govuk_link_to "Find my URN", "#", class: "govuk-link" %></p>
       <%= f.govuk_email_field(
             :email,
-            label: { text: "Email address" },
-            hint: { text: "This should be the email address provided by your school" },
-            autocomplete: "email")
+            value: @email,
+            label: { text: "Your school email address", class: "govuk-label govuk-label--s" })
       %>
 
-      <%= f.govuk_submit "Sign up" %>
+      <%= f.govuk_submit "Continue" %>
     <% end %>
 
   </div>
```
This file and `users/registrations/new` look remarkably similar. Can we commonize them a little? The form for example.

amazimbe replied:
I can't see anything we can commonise and I think trying to do so for things which are likely to need frequent changes like these two would probably not be a good idea.

kicferk1 replied:
Well, the form is similar.
We have 
```
    <%= form_for(:induction_coordinator_profile, url: users_check_details_path) do |f| %>
      <%= f.govuk_text_field(
            :school_urn,
            value: @school&.urn,
            hint: { text: "6 digit number", class: "govuk-hint" },
            label: { text: "Your school's Unique Reference Number (URN)", class: "govuk-label govuk-label--s" })
      %>
      <p class="govuk-body govuk-!-font-weight-bold">
        <%# TODO replace with actual link when available %>
        <%= govuk_link_to "Find my URN", "#", class: "govuk-link" %>
    </p>
      <%= f.govuk_email_field(
            :email,
            value: @user&.email,
            autocomplete: "email",
            label: { text: "Your school email address", class: "govuk-label govuk-label--s" })
      %>
      <%= f.govuk_submit "Continue" %>
    <% end %>
```
in the other one. The only differences I can spot are that in `registrations/new`, we are passing urn and email through other objects (user and school), while here they are their own variables.

Well, that and one has a TODO on a dead link, and the other doesn't. 

If you think they will very likely change and it is just a coincidence they look the same I'm happy to let them be coincidentally similar though :)

kicferk1 commented on app/views/users/registrations/verification_email_sent.html.erb:
```diff
@@ -0,0 +1,10 @@
+  <div class="govuk-grid-column-two-thirds">
+    <h2 class="govuk-heading-l">Verify your email</h2>
+    <p>We have sent you a link to verify your email. This will expire in 24 hours.</p>
+    <p>It may be in your spam or junk folder</p>
+    <br>
```
There are govuk design classes for spacing, we should use them over brs. 

https://design-system.service.gov.uk/styles/spacing/

`govuk-!-margin-bottom-2` or something like that will do I think.

kicferk1 commented on app/views/users/sessions/redirect_from_magic_link.html.erb:
```diff
@@ -1,7 +1,7 @@
   <div class="govuk-grid-column-two-thirds">
-    <h2 class="govuk-heading-l">Continue to your account</h2>
-
+    <h3 class="govuk-heading-l">Sign in successful</h3>
```
Damn, we will need to sort out our headers I think. (as in h1, h2, h3, and where to use them to be semantically correct :P )
I would keep using `h2` here though.

amazimbe replied:
The wireframes show different sizes so I just picked the sizes that seemed to match the designs better. But yes, we may need to revisit this for consistency and to be semantically correct. I'll probably also get some feedback from the designer.

kicferk1 commented on app/webpacker/styles/registration.scss:
```diff
@@ -0,0 +1,7 @@
+.confirmed-banner {
+  background-color: #00703c;
+  padding: 50px 30px;
+  margin: 15px;
+  text-align: center;
+  color: govuk-colour("white") !important;
+}
```
I imagine you took the styles from a site like https://design-system.service.gov.uk/patterns/confirmation-pages/

I guess that means we don't have them in our govuk styling by default?

amazimbe replied:
No, I just made the styles and wasn't aware of  https://design-system.service.gov.uk/patterns/confirmation-pages/. I'll try and use the classes from there as we probably already have the styles.

kicferk1 commented on spec/models/school_spec.rb:
```diff
@@ -15,6 +15,72 @@
+      let(:user) { create(:user, confirmed_at: nil) }
+      let!(:coordinator) { create(:induction_coordinator_profile, user: user, schools: [school]) }
+
+      it "returns false if no one has registered a school" do
+        expect(school.partially_registered?).to be true
+      end
```
Eh, the test description doesn't match what it is checking.

kicferk1 commented on spec/requests/users/registrations/check_details_spec.rb:
```diff
@@ -0,0 +1,90 @@
+  let(:email) { Faker::Internet.email(domain: school.domains.first) }
+
+  describe "POST /users/check_email" do
+    it "renders correct template when no schools are found" do
```
Rather than 'correct template' in many places in this file, I would expect the names of templates in the test description - so that when someone has a failing test they won't be tempted to just quickly change the template in it.

kicferk1 reviewed (commented):
Looks pretty good overall, most comments are minor or questions :)

ethanmills commented:
Let's make use of the review app here 🙂 Would be really good to put together a set of input parameters to trigger different screen and send them over to Al and Shani and check they're happy

amazimbe commented:
> Let's make use of the review app here 🙂 Would be really good to put together a set of input parameters to trigger different screen and send them over to Al and Shani and check they're happy

@ethanmills I was thinking more along the lines of doing a quick demo with them so they can give me some feedback but probably after merging the pr. I could then create smaller tickets to address any recommendations they make.

cfq commented:
@amazimbe @ethanmills I think a preview app is a good idea. It's why we set them up in the first place – to be able to demo before merging.

amazimbe commented:
> @amazimbe @ethanmills I think a preview app is a good idea. It's why we set them up in the first place – to be able to demo before merging.

Agreed @ethanmills @cfq

ethanmills commented on app/models/school.rb:
```diff
@@ -17,4 +17,24 @@ def full_address
+  def partially_registered?
+    return false if not_registered?
+
+    induction_coordinator_profiles
```
This is just `!fully_registered?`, right?

amazimbe replied:
No, I think there's more to it than just !fully_registered? especially as we have 3 states i.e. not_registered? as well.

ethanmills replied:
Sorry, specifically lines 35-38

ethanmills replied:
The whole function is just `!(fully_registered? || not_registered?)` I believe

amazimbe replied:
Probably! More than happy to simplify it.

ethanmills commented on app/views/users/registrations/new.html.erb:
```diff
@@ -1,19 +1,25 @@
+            label: { text: "Your school's Unique Reference Number (URN)", class: "govuk-label govuk-label--s" })
+      %>
+      <p class="govuk-body govuk-!-font-weight-bold">
+        <%= govuk_link_to "Find my URN", "#", class: "govuk-link" %>
```
Would be good to have a todo here so it doesn't get lost

amazimbe replied:
There was some talk of removing that altogether in the demo this morning. But yeah, I can add a todo.

ethanmills commented:
Noticed a couple of things. Tried logging in to the review app (with no account). Saw this 
![image](https://user-images.githubusercontent.com/40488007/105871371-235d3480-5ff1-11eb-9d32-e389267db9ed.png)
then clicking on the button showed me this
![image](https://user-images.githubusercontent.com/40488007/105871404-2c4e0600-5ff1-11eb-90b9-861292fe8cff.png)

amazimbe commented:
> Noticed a couple of things. Tried logging in to the review app (with no account). Saw this
> ![image](https://user-images.githubusercontent.com/40488007/105871371-235d3480-5ff1-11eb-9d32-e389267db9ed.png)
> then clicking on the button showed me this
> ![image](https://user-images.githubusercontent.com/40488007/105871404-2c4e0600-5ff1-11eb-90b9-861292fe8cff.png)

@ethanmills that was in the design. I went through it with Al and Shani this morning and that's one of the things they will be redesigning.

ethanmills commented on app/views/users/sessions/email_not_found.html.erb:
```diff
@@ -1,7 +1,7 @@
-    <%= govuk_link_to "Register your school", users_check_email_path %>
+    <h2 class="govuk-heading-l">Your school has not registered</h2>
+    <p>You must be responsible for running your induction programme to register your school.</p>
+    <%= govuk_link_to "Register your school", users_information_path, button: true %>
```
Is this not meant to be `users_new_path`?

amazimbe replied:
No, first to those information pages you had in your screenshots earlier

ethanmills replied:
Are you working from these designs? 
![image](https://user-images.githubusercontent.com/40488007/106033934-3132b880-60ca-11eb-9e20-bc0dc94cd123.png)

ethanmills replied:
Which design are you working from?

amazimbe replied:
https://miro.com/app/board/o9J_lZ2k8oc=/

amazimbe replied:
Al and Shani have been updating it so it's slightly diff from what it was. I'll have to update it later when they are done.

ethanmills commented on app/views/users/confirmations/confirmed.html.erb:
```diff
@@ -0,0 +1,8 @@
+<div class="govuk-grid-row">
+  <h2 class="govuk-heading-m confirmed-banner">Your school is registered</h2>
```
This wants to be `<%= govuk_panel title: "Your school is registered", body: nil %>`

amazimbe replied:
@ethanmills does this apply to all the page titles?

ethanmills replied:
No, this gives you a [panel](https://design-system.service.gov.uk/components/panel/)

amazimbe commented:
@ethanmills @kicferk1 could you take a final look please. I have addressed all the feedback for now and any further design changes from Al and Shani will be implemented in separate prs.

kicferk1 commented on app/webpacker/styles/registration.scss:
```diff
@@ -0,0 +1,3 @@
+.confirmed-banner {
+  margin: 15px;
```
Nice :)

amazimbe replied:
We can have a chat when I implement updates to the design but yes, I think it's better to keep these separate rather than having a form with conditionals and all that.
